.github/workflows/ci.yaml

```
@@ -79,6 +79,114 @@ jobs:
+      - name: Cache container layers
+        uses: actions/cache@v3
+        with:
+          path: /tmp/.buildx-cache
+          key: ${{ runner.os }}-buildx-nox-${{ matrix.arch }}-${{ matrix.os }}-${{ matrix.python }}
+
+      - name: Build image
+        uses: docker/build-push-action@v4
+        with:
+          context: .github/containers/nox-cross-arch
+          file: .github/containers/nox-cross-arch/${{ matrix.arch }}-${{ matrix.os }}-python-${{ matrix.python }}.Dockerfile
+          platforms: linux/${{ matrix.arch }}
+          tags: localhost/nox-cross-arch:latest
+          push: false
+          load: true
+          cache-from: type=local,src=/tmp/.buildx-cache
+          cache-to: type=local,dest=/tmp/.buildx-cache-new,mode=max
+
+      # Refer to the workaround mentioned above
+      - name: Move cache
+        run: |
+          rm -rf /tmp/.buildx-cache
+          mv /tmp/.buildx-cache-new /tmp/.buildx-cache
+
+      # Cache pip downloads
+      - name: Cache pip downloads
+        uses: actions/cache@v3
+        with:
+          path: /tmp/pip-cache
+          key: nox-${{ matrix.nox-session }}-${{ matrix.arch }}-${{ matrix.os }}-${{ matrix.python }}-${{ hashFiles('**/pyproject.toml') }}
+
+      # This ensures that the docker container has access to the pip cache.
+      # Changing the user in the docker-run step causes it to fail due to
+      # incorrect permissions. Setting the ownership of the pip cache to root
+      # before running is a workaround to this issue.
+      - name: Set pip cache owners to root for docker
+        run: if [[ -e /tmp/pip-cache ]]; then sudo chown -R root:root /tmp/pip-cache; fi
+
+      - name: Run nox
+        run: |
+          docker run \
+            --rm \
+            -v $(pwd):/${{ github.workspace }} \
+            -v /tmp/pip-cache:/root/.cache/pip \
+            -w ${{ github.workspace }} \
+            --net=host \
+            --platform linux/${{ matrix.arch }} \
+            localhost/nox-cross-arch:latest \
+            bash -c "pip install -e .[dev-noxfile]; nox --install-only -e ${{ matrix.nox-session }}; pip freeze; nox -e ${{ matrix.nox-session }}"
+        timeout-minutes: 30
+
+      # This ensures that the runner has access to the pip cache.
+      - name: Reset pip cache ownership
+        if: always()
+        run: sudo chown -R $USER:$USER /tmp/pip-cache
+
+  # This job runs if `nox-cross-arch` ran and succeeded.
+  # This is required because, when the `nox-cross-arch` job is skipped, its
+  # inner matrix is not expanded, and branch protection rules on the
+  # inner-matrix jobs get stuck. So instead of `nox-cross-arch`, this job can be
+  # used in the branch protection rules.
+  # At the time of writing this, there is an ongoing discussion about this here:
+  # https://github.com/orgs/community/discussions/9141
+  nox-cross-arch-all:
+    name: A dummy non-matrix job that only passes if the nox-cross-arch job passed.
```

I would name this exactly the same as `nox-cross-arch`, so the hack gets hidden to the users, the branch protecting rules work on the name, and the name can be duplicated (we use that trick to make DCO work with merge queues).

```suggestion
    name: Cross-arch tests with nox
```

---

Sure, done.